File: 0020_Valid_Parentheses/valid_parentheses.cpp
// Check py file for explaining comments

class Solution {
public:
    bool isValid(string s) {
        vector<char> stack;
        unordered_map<char, char> paras = {
            {')' , '('},
            {'}' , '{'},
            {']' , '['},
        };

        for (int i = 0; i < s.size(); i++){
            char ch = s[i];
            if(ch == '(' || ch == '{' || ch == '['){
                stack.push_back(ch);
            }
            else{
                if(stack.empty() || paras[ch] != stack.back())
                    return false;
                stack.pop_back();
            }
        }
        if (stack.empty())
            return true;
        else
            return false;
    }
};

File: 0020_Valid_Parentheses/Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        paras = { ")":"(", "}":"{", "]":"["} # I will be asking for opening brackets so those will be in values (check in line 10), and so closes will be in keys.
        for ch in s: # For open brackets
            if ch in paras.values():
                stack.append(ch)

            else: # For close brackets
                if not stack or stack[-1] != paras[ch]:
                    # Not stack means Stack khali? yes!
                    # open na match korle true debe
                    # total condition true hole False trigger korbe 
                    return False
                stack.pop()

        if not stack:
            # Stack khali, tarmane extra kichu pore nei tai true
            return True
        else:
            return False